### DISCUSSION TITLE
Why can't getStaticProps include the "req" object when not executed during build

### DISCUSSION THREAD

**Discussion by wmattei** [vercel/next.js#21431] [Help] [answered]
I understand that when `getStaticProps` is ran during the build time we cannot have the `req` object, as there isn't any request being made, but what if there  is? Like on a cache revalidate request, where the user receives the old content while the new one is being regenerated, (_stale-while-regenerate_).

In this case, the `getStaticProps` function is being triggered by a request, so why can't i have the `req` object?

**Comment by FSaldanha** [comment]
Have you already checked this? #17269

**Comment by bennettdams** [comment]
Imagine the possibilities: You could create static sites which are filled with data from a DB, but is only visible to authenticated users.
We as developers could skip client-side fetching and server-side rendering (without complicated caching), but just create static pages with fresh data every x seconds -> "Incremental Static Regeneration" in the background.
As this data is only shown to authenticated users, having the request in `getStaticProps` (or maybe another, new handler function next to it?) is a must (e.g. to check cookies).

I have described the two requirements in a little more detail here:  https://github.com/vercel/next.js/discussions/28959

**Reply by lucasbasquerotto** [reply]
Only the 1st time the cookie check validation would be done (inside the page) though, because in the next accesses it would return the static page that was generated. I assume such validation should be done before reaching the page, otherwise it could even be insecure (if some authentication data from a user is different from the other). 

You could create 2 rewrite rules based on the cookie value (or existence), in one case leading to one page (authenticated) and in the other case leading to another (unauthenticated), even tough the url is the same (because it's a rewrite, not a redirect).

**Reply by bennettdams** [reply]
>because in the next accesses it would return the static page that was generated

@lucasbasquerotto It doesn't have to be that way though, right? Couldn't Next.js just give us another handler (like a function next to `getStaticProps`) that is always executed when a route for this static site is accessed? That way you could always redirect if the user is not authenticated.

**Reply by FSaldanha** [reply]
Couldn't you achieve this using Next.js built-in [custom server](https://nextjs.org/docs/advanced-features/custom-server) and a third-party auth provider?

[This tutorial](https://auth0.com/blog/ultimate-guide-nextjs-authentication-auth0/#The-Custom-Server-Approach) shows an example using Express.js along Next.js, but I think you could tweak it to use Next.js only.

**Reply by bennettdams** [reply]
@FSaldanha You probably can, but why add extra complexity if this could "easily" be a feature that comes with Next.js out of the box? A custom server brings so much more problems regarding maintenance, caching, deployment, error handling, etc.
The solution discussed here looks "easy" to implement as an outsider, so it would be interesting if a) the implementation needed from the Next.js team is way bigger than how it looks to an outsider like us or b) the Next.js team just didn't think this is a feature us developers need.

**Comment by bennettdams** [comment, marked as answer]
With Next.js 12, this is now possible out of the box with [middlewares](https://nextjs.org/blog/next-12#introducing-middleware). You have access to the request (e.g. for cookies) and can redirect, before the static site was sent to the user.

**Reply by Seba-Bloque** [reply]
And an example of how to use it? Please?
